File: cpp/StrategyDemo.cpp
#include <map>
#include <iostream>
#include <string>
#include <algorithm>

using namespace std;

//author sbkun
class Discount {
	public:
		virtual ~Discount() {}
		virtual float calculate(float price) = 0;
};

class Vip: public Discount {
public:
	float calculate(float price) override {
		return price*0.8;
	}
};

class GoldVip: public Discount {
public:
	float calculate(float price) override {
		return price*0.5;
	}
};

class DiscountFactory {
public:
	static DiscountFactory& getInstance() {
		static DiscountFactory instance;
		return instance;		
	}

	Discount* getDiscount(const string& vipType) {
		string t = vipType;
		std::transform(t.begin(), t.end(), t.begin(), ::tolower);
		return discounts_[t];
	}

private:
	DiscountFactory() {
		discounts_["vip"] = new Vip;
		discounts_["goldvip"] = new GoldVip;
	}

	map<string, Discount*> discounts_;
};

class Product {
public:
	Product(float price): price_(price) { }
	float getPrice(const string& vipType) {
		Discount* discout = DiscountFactory::getInstance().getDiscount(vipType);
		if (discout != NULL)	{
			return discout->calculate(price_);
		}	
		return price_;
	}

private:
	float price_;
};


int main(int argc, char* argv[]) {
	Product product(100);
	cout<<product.getPrice("vip")<<endl;
	cout<<product.getPrice("goldvip")<<endl;
	cout<<product.getPrice("haha")<<endl;
	return 0;
}

File: java/behavioral/StrategyDemo.java
package behavioral;

import java.util.HashMap;
import java.util.Map;

interface Discount {
    double calculate(double price);
}

class VipDiscount implements Discount {
    //会员8折
    @Override
    public double calculate(double price) {
        return price * 0.8;
    }
}

class GoldVip implements Discount {
    //黄金会员5折
    @Override
    public double calculate(double price) {
        return price * 0.5;
    }
}

class DiscountFactory {
    static{
        System.out.println("haha");
    }
    private static final DiscountFactory INSTANCE = new DiscountFactory();

    public static DiscountFactory getInstance() {
        return INSTANCE;
    }

    private DiscountFactory() {
        System.out.println(this.getClass().getSimpleName() + " init");
        discounts.put("vip", new VipDiscount());
        discounts.put("goldvip", new GoldVip());

    }

    Discount getDiscount(String vipType) {
        String t = vipType.toLowerCase();
        return discounts.get(t);
    }

    private Map<String, Discount> discounts = new HashMap<>();
}

class Product {
    Product(double original_price) {
        this.original_price = original_price;
    }

    public double getPrice(String vipType) {
        Discount discount = DiscountFactory.getInstance().getDiscount(vipType);
        if (discount != null) {
            return discount.calculate(original_price);
        }
        return original_price;
    }

    private double original_price;
}

public class StrategyDemo {
    public static void main(String args[]) {
        System.out.println("in main");
        Product product = new Product(100);
        System.out.println(product.getPrice("goldvip"));
        System.out.println(product.getPrice("vip"));
        System.out.println(product.getPrice("haha"));
    }

}


